server/server.go

```diff
@@ -1737,15 +1737,26 @@ func (s *Server) watchServicePrimaryAddrLoop(serviceName string) {
 	} else {
 		log.Warn("service primary addr doesn't exist", zap.String("service-key", serviceKey))
 	}
+	watcher := clientv3.NewWatcher(s.client)
+	defer watcher.Close()
 
-	watchChan := s.client.Watch(ctx, serviceKey, clientv3.WithPrefix(), clientv3.WithRev(rev))
 	for {
-		select {
-		case <-ctx.Done():
-			log.Info("server is closed, exist watch service primary addr loop", zap.String("service", serviceName))
-			return
-		case res := <-watchChan:
-			for _, event := range res.Events {
+		watchChan := watcher.Watch(ctx, serviceKey, clientv3.WithPrefix(), clientv3.WithRev(revision))
+		for wresp := range watchChan {
+			if wresp.CompactRevision != 0 {
+				log.Warn("required revision has been compacted, use the compact revision",
+					zap.Int64("required-revision", revision),
+					zap.Int64("compact-revision", wresp.CompactRevision))
+				revision = wresp.CompactRevision
+				break
+			}
+			if wresp.Canceled {
```

Comment:
It seems that, you haven't forgotten your comment in my pr and didn't check my reply :)

At the same place, you asked "Do we need to solve other errors?", I replied:

I checked the code in etcd watch.go. There is no other errors to solve, but I do think here using wrep.Err() != nil is better than using wresp.Canceled, because the former can accommodate future etcd changes.

Etcd Watch() returns three types of errors -- closed error, compacted error and cancelled error. closed error is captured by range watchChan, compacted error is captured by wresp.CompactRevision != 0, and cancelled error is captured here. Etcd Watch() will retry Watch() for all other errors. Using wrep.Err() != nil is better than using wresp.Canceled.

Changed to using wrep.Err() != nil here.

Member
Author
@[binshi-bing](https://github.com/binshi-bing) binshi-bing Pending
I checked the code in etcd watch.go. There is no other errors to solve, but I do think here using wrep.Err() != nil is better than using wresp.Canceled, because the former can accommodate future etcd changes.

Etcd Watch() returns three types of errors -- closed error, compacted error and cancelled error. closed error is captured by range watchChan, compacted error is captured by wresp.CompactRevision != 0, and cancelled error is captured here. Etcd Watch() will retry Watch() for all other errors. Using wrep.Err() != nil is better than using wresp.Canceled.

Changed to using wrep.Err() != nil here.

Reply:
sorry, I forgot about it when the copilot finished